asg/parallel-graph: Update assignment (#30)
cs-pub-ro/operating-systems: content/assignments/parallel-graph/README.md

# Parallel Graph

## Objectives

- Learn how to design and implement parallel programs
- Gain skills in using synchronization primitives for parallel programs
- Get a better understanding of the POSIX threading and synchronization API
- Gain insight on the differences between serial and parallel programs

## Statement

Implement a generic thread pool, then use it to traverse a graph and compute the sum of the elements contained by the nodes.
You will be provided with a serial implementation of the graph traversal and with most of the data structures needed to implement the thread pool.
Your job is to write the thread pool routines and then use the thread pool to traverse the graph.

## Support Code

The support code consists of the directories:

- `src/` is the skeleton parallel graph implementation.
  You will have to implement missing parts marked as `TODO` items.

- `utils/` utility files (used for debugging & logging)

- `tests/` are tests used to validate (and grade) the assignment.

## Implementation

### Thread Pool Description

> A thread pool contains a given number of active threads that simply wait to be given specific tasks.
> The threads are created when the thread pool is created they poll a task queue until a task is available.
> Once tasks are put in the task queue, the threads start running the task.
> A thread pool creates N threads when the thread pool is created and does not destroy (join) them throughout the lifetime of the thread pool.

Reviewer: ```suggestion
A thread pool contains a given number of active threads that simply wait to be given specific tasks.
The threads are created when the thread pool is created.
Each thread continuously polls the task queue for available tasks.
Once tasks are put in the task queue, the threads poll tasks, and start running them.
A thread pool creates **N** threads upon its creation and does not destroy (join) them throughout its lifetime.
```

That way, the penalty of creating and destroying threads ad-hoc is avoided.
As such, you must implement the following functions (marked with `TODO` in the provided skeleton, in `src/os_threadpool.c`):

- `enqueue_task()`: Enqueue task to the shared task queue.
  Use synchronization.
- `dequeue_task()`: Dequeue task from the shared task queue.
  Use synchronization.
- `wait_for_completion()`: Wait for all worker threads.
  Use synchronization.
- `create_threadpool()`: Create a new thread pool.
- `destroy_threadpool()`: Destroy a thread pool.
  Assume all threads have been joined.

You must also update the `os_threadpool_t` structure in `src/os_threadpool.h` with the required bits for synchronizing the parallel implementation.

Notice that the thread pool is completely independent of any given application.
Any function can be registered in the task queue.

> Since the threads are polling the task queue indefinitely, you need to find a condition for the threads to stop once the graph has been traversed completely.

Reviewer: ```suggestion
Since the threads are polling the task queue indefinitely, you need to define a condition for them to stop once the graph has been traversed completely.
```

That is, the condition used by the `wait_for_completion()` function.

> The recommended way is to note when no threads has any more work to do.

Reviewer: ```suggestion
The recommended way is to note when no threads have any more work to do.
```

> Since no thread is doing any work, there is also no other work to generate.

Reviewer: ```suggestion
Since no thread is doing any work, no other task will be created.
```

Total:                                                              90/100
```

### Running the Linters

To run the linters, use the `make lint` command in the `tests/` directory:

```console
student@so:~/.../assignments/parallel-graph/tests$ make lint
[...]
cd .. && checkpatch.pl -f checker/*.sh tests/*.sh
[...]
cd .. && cpplint --recursive src/ tests/ checker/
[...]
cd .. && shellcheck checker/*.sh tests/*.sh
```

Note that the linters have to be installed on your system: [`checkpatch.pl`](https://.com/torvalds/linux/blob/master/scripts/checkpatch.pl), [`cpplint`](https://github.com/cpplint/cpplint), [`shellcheck`](https://www.shellcheck.net/).
They also need to have certain configuration options.

> It's easiest to run them in a Docker-based setup with everything configured:

Reviewer: ```suggestion
It's easiest to run them in a Docker-based setup with everything configured.
```

### Fine-Grained Testing

Input tests cases are located in `tests/in/`.

> If you want manually run a single test, use commands such as below while in the `src/` directory:

Reviewer: ```suggestion
If you want to run a single test, use commands such as below while in the `src/` directory:
```
